### ISSUE TITLE
Fluent binding ergonomics

### ISSUE THREAD

**Issue by dginev** [dginev/latexml-oxide#24]
The just merged #23 has fully set the stage to attack the `Def*` mini language with a goal of full parity with the latexml `.ltxml` definitions. 

It is now possible to hide `state` entirely, AND to invoke nested binding constructs, passing a stomach where needed.

The steps remaining that I can see:
 * [x] auto-bind the closest state in closures such as `before_digest`
 * [x] refactor existing bindings to never use the explicit state form
 * ~~remove the explicit state form entirely, in favor of using local scope redefinitions of the `state!()` macro, which are less prone to human error, and will simplify `setup.rs` significantly.~~
 * [x] compile-time upgrade: move the heavy load definitions into dedicated rust functions, such as the existing `fn def_macro`. 
 * [x] Biggest remaining pain-point in writing down definitions: there is a need for better argument parsers for dependent closures. In particular, it would be great if `before_digest => before_digest!(...)` could become simply `before_digest!(...)`

**Comment by dginev** [comment]
The `bind_state!(st)` technique introduced in #25 gets us most of the way to the auto-bind behaviour, but there is still an open question of how to pass in the "inner"/"outer" flag to the macro generator.

**Comment by dginev** [comment]
The `bind_state!(st)` technique has now reached its ultimate form, and we can invoke "stateless" macro calls with the correct state being auto-bound by the rtx machinery. Details at #27

**Comment by dginev** [comment]
The explicit `state` argument of bindings should remain available however, as nesting 2 levels or deeper requires an explicit state argument to be provided, at the moment. It's rare, but good to have.

**Comment by dginev** [comment]
I ended up succeeding in simplifying something not on the original list:
```
sub[stomach, args, state] {...}
```

can  now be written as simply `sub {...}` when you don't need to use the three input arguments, e.g. when returning some constants, or doing a side-effect that only requires state. The *crucial* feature that made this possible is that we now have auto-binding of the inner state, so it's perfectly easy to write something like:
```
sub { AssignValue!("inPreamble", false); }
```

without ever thinking of state explicitly. This is getting near the Perl way of thinking, but it's just a syntactic slight of hand - the expanded code does care very strictly about which state is being used and how, performs mutability and borrowing checks etc.

**Comment by dginev** [comment]
Here is a more complete example of the most recent binding I've ported:

LaTeXML original:
```perl
DefEnvironment('{itemize}',
  "<ltx:itemize xml:id='#id'>#body</ltx:itemize>",
  properties => sub { beginItemize('itemize', '@item'); },
  beforeDigestEnd => sub { Digest('\par'); },
  locked => 1, mode => 'text');
```

Current Rust port:
```rust
  DefEnvironment!("{itemize}",
    "<ltx:itemize xml:id='#id'>#body</ltx:itemize>",
    properties => properties!(sub { BeginItemize!("itemize", "@item") }),
    before_digest_end => before_digest!(sub { Digest!("\\par")?; }),
    locked => true, mode => "text".into_option());
```

Desired Rust goal:
```rust
  DefEnvironment!("{itemize}",
    "<ltx:itemize xml:id='#id'>#body</ltx:itemize>",
    properties => { BeginItemize!("itemize", "@item") },
    before_digest_end => { Digest!("\\par")?; },
    locked => true,
    mode => "text"
  );
```

The blocker since the beginning has been that if I want to allow flexible argument lists via `=>`, I can not use any macro patterns in that iterator of `$($key:ident=>$val:expr),*`. The biggest reason I can't just use the `{}` blocks in this example is that there is an implicit closure signature with the various input variables (think `state: &mut State`), which need to be filled in by the macro machinery. And if we ever explicitly use a variable, we have to provide it a priori, so that the block can be bound to it. An unfortunate example would be:
```rust
    properties => properties!(sub[stomach, args, state] { ... })
```
which I would hope to ideally reduce to:
```rust
    properties => sub[stomach, args, state] { ... }
```

The rust folks on Discord had suggested that I can escape most/all of these constraints by writing a custom parser as a special construct in `rtx_codegen`, and maybe that is something to explore... 

It's certainly an enhancement which will make porting the rest of LaTeXML *much* easier, as the distance to the direct copy-paste becomes very minimal.

Here is a link to one [relevant discussion](https://users.rust-lang.org/t/create-a-struct-from-macro-rules/19829/2) of this type of macro `munching`.

**Comment by dginev** [comment]
Oh nice, the `TT munching` technique actually worked pretty directly and successfully. Linking to [another tutorial](https://danielkeep.github.io/tlborm/book/pat-incremental-tt-munchers.html) on the subject.

I'm starting a big refactor (= simplification) of all `Def*` macros and their uses, hoping this all goes swimmingly.

**Comment by dginev** [comment]
Solved via #46 , and I am delighted I discovered the token tree munchers!
